[FLINK-19465][runtime / statebackends] Add CheckpointStorage interface and wire through runtime

sjwiesman opened the pull request:
## What is the purpose of the change

This adds the new `CheckpointStorage` interface as described in FLIP-142 and wires it through the runtime. Public API changes will be added in a follow-up PR. 

## Brief changelog

675d93c  Adds the new `CheckpointStorage` interface, factory, and loader.

6a41e53    Wires the `CheckpointStorage` object through the StreamGraph, ExecutionGraph, StreamTask, and CheckpointCoordinator

550bb03   Removes the checkpoint-storage related methods from `StateBackend` and adds `CheckpointStorage` as an interface to the existing state backend implementations. This commit then makes misc fixes to tests so everything passes. 

## Verifying this change

UT tests and also existing E2E tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

sjwiesman commented:
Note to reviewers, I decided not update the javadoc for StateBackend in this PR. I'll update it in the follow up when I add the new StateBackend implementations.

sjwiesman commented on flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/utils/ExecutorUtils.java:
```diff
@@ -79,6 +79,7 @@ public static void setBatchProperties(StreamGraph streamGraph, TableConfig table
 		streamGraph.setAllVerticesInSameSlotSharingGroupByDefault(false);
 		streamGraph.setScheduleMode(ScheduleMode.LAZY_FROM_SOURCES_WITH_BATCH_SLOT_REQUEST);
 		streamGraph.setStateBackend(null);
+		streamGraph.setCheckpointStorage(null);
```
@aljoscha I'm pretty certain this is correct but could you please double-check. I'm also not sure if I need to explicitly do something for batch execution somewhere else now that it's supported for DataStream.

sjwiesman commented:
Fixed the missing license, sorry.

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java:
```diff
@@ -328,6 +329,7 @@ private void setBatchStateBackendAndTimerService(StreamGraph graph) {
         if (useStateBackend) {
             LOG.debug("Using BATCH execution state backend and timer service.");
             graph.setStateBackend(new BatchExecutionStateBackend());
+            graph.setCheckpointStorage(new BatchExecutionCheckpointStorage());
```
Should there be a similar call for other cases (line 303)?

sjwiesman replied:
This was in the follow up PR. This is the weird space where we are getting close the public API. I have backported `graph.setCheckpointStorage` on line 303ish but not setting the instance variable `checkpointStorage` that I have left for the follow up because it involves modifying `StreamExecutionEnvironment`.

rkhachatryan commented on flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBStateBackend.java:
```diff
@@ -468,12 +474,12 @@ private File getNextStoragePath() {
     @Override
     public CompletedCheckpointStorageLocation resolveCheckpoint(String pointer) throws IOException {
-        return checkpointStreamBackend.resolveCheckpoint(pointer);
+        return ((CheckpointStorage) checkpointStreamBackend).resolveCheckpoint(pointer);
```
It seems possible to change the field type to `CheckpointStorage` and remove these casts.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/checkpoints/CheckpointConfigInfo.java:
```diff
@@ -79,6 +81,9 @@
     @JsonProperty(FIELD_NAME_STATE_BACKEND)
     private final String stateBackend;
 
+    @JsonProperty(FIELD_NAME_CHECKPOINT_STORAGE)
+    private final String checkpointStorage;
```
I couldn't find rendering of this field in the UI. 
Or is it implemented in the subsequent PR?

sjwiesman replied:
Forgot to backport, nice catch.

sjwiesman replied:
I didn't add it to the UI because the state_backend field is also not rendered in the UI. What do you think?

rkhachatryan replied:
Hmm..you're right. I think it would be nice to show it but can be implemented in a separate PR/ticket.

sjwiesman replied:
Agreed. I will open a follow-up ticket.

rkhachatryan reviewed (changes requested):
Thanks for this PR @sjwiesman 
I've left some comments, PTAL.

sjwiesman commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStorageLoader.java:
```diff
@@ -0,0 +1,180 @@
+import java.io.IOException;
+
+/** This class contains utility methods to load checkpoint storage from configurations. */
+public class CheckpointStorageLoader {
```
It should actually be internal, nice catch

sjwiesman commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStorageLoader.java:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    public static CheckpointStorage fromApplicationOrConfigOrDefault(
+            @Nullable CheckpointStorage fromApplication,
```
Flink's style guide says not to use `Optional` for function arguments. 
https://flink.apache.org/contributing/code-style-and-quality-java.html#java-optional

rkhachatryan replied:
Do you mean
> Do not use Optional as a function argument, instead either overload the method or use the Builder pattern for the set of function arguments.

I think the intention is to avoid arguments that can be empty/null. According to it the method should be refactored.

But I think the very purpose of this method is to chose which arguments to use. And for that I'd use `Optional`.

sjwiesman commented on flink-runtime/src/test/java/org/apache/flink/runtime/state/CheckpointStorageLoadingTest.java:
```diff
@@ -0,0 +1,228 @@
+        @Override
+        public CheckpointStorage createFromConfig(ReadableConfig config, ClassLoader classLoader)
+                throws IllegalConfigurationException, IOException {
+            throw new IOException("fail!");
```
I dropped the IOException. That's a leftover from an earlier iteration I missed when cleaning up the PR.

sjwiesman commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStorageLoader.java:
```diff
@@ -0,0 +1,180 @@
+        Preconditions.checkNotNull(classLoader, "classLoader");
+
+        final String storageName = config.get(CheckpointingOptions.CHECKPOINT_STORAGE);
+        if (storageName == null) {
+            return null;
```
This is required to maintain proper backward compatibility with existing flink-confs. While we cannot make this config strictly required I will add some warning level logging encouraging users to add it. 

Also, +1 on the Optional.

sjwiesman commented:
@rkhachatryan appreciate the speedy review. I hope I've addressed all your comments. Please take a look and let me know if you have any other concerns.

rkhachatryan reviewed (approved):
Thanks for updating the PR @sjwiesman !
LGTM

sjwiesman commented:
Thanks! Will merge and rebase the follow-up later today.

sjwiesman commented:
Merged in 2426d84268ad330ec661a0ea6bedb2a3a2bd08e6

curcur commented:
Hoops, I commented before I noticed this PR is closed.

There are a couple of places that "state-backed" should be replaced by "checkpoint storage", but we can fix them in the follow-up PRs I guess.
